### PR TITLE
CopyObject: Set Content-Type to application/octet-stream if it is not set

### PR DESCRIPTION
**Pull request by kannappanr**
## Description
Set Content-Type by default to application/octet-stream, if the client does not set it.
## Motivation and Context
In the case of Gateway, Azure and GCS set default content-type to '' and AWS sets it to `binary/octet-stream`. 
There will be a consistent default content-type.

## Regression
Not sure

## How Has This Been Tested?
Run this fix in `Gateway S3` mode with `mint` and `minio-py`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#1231b017-f201-446b-9d29-9ec833fa446f) |
| mint-xl.sh | :x: [more...](#8946fc9c-b1fb-4d6a-b8b9-1ff51487fc39) |
| mint-large-bucket.sh | :x: [more...](#3e20cfe2-8222-4414-b953-6fa52d7798ec) |
<a name="8946fc9c-b1fb-4d6a-b8b9-1ff51487fc39"></a>
### 6958-d366ec5/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31202
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a355d2da40a7:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 13 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 31 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 46 seconds
(8/13) Running minio-go tests ... done in 42 seconds
(9/13) Running minio-java tests ... FAILED in 2 minutes and 34 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 549,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```
<a name="1231b017-f201-446b-9d29-9ec833fa446f"></a>
### 6958-d366ec5/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30171
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cc69ae368bc8:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 13 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 33 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 43 seconds
(8/13) Running minio-go tests ... done in 40 seconds
(9/13) Running minio-java tests ... FAILED in 2 minutes and 31 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 777,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```
<a name="3e20cfe2-8222-4414-b953-6fa52d7798ec"></a>
### 6958-d366ec5/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32576
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 9153080b871f:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 4 seconds
(2/13) Running aws-sdk-go tests ... done in 4 seconds
(3/13) Running aws-sdk-java tests ... done in 8 seconds
(4/13) Running aws-sdk-php tests ... done in 58 seconds
(5/13) Running aws-sdk-ruby tests ... done in 32 seconds
(6/13) Running mc tests ... done in 1 minutes and 30 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 31 seconds
(8/13) Running minio-go tests ... done in 1 minutes and 55 seconds
(9/13) Running minio-java tests ... FAILED in 3 minutes and 41 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 300,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```

File: cmd/object-handlers.go [hunk 1]
```diff
@@ -976,6 +976,11 @@ func (api objectAPIHandlers) CopyObjectHandler(w http.ResponseWriter, r *http.Re
 		srcInfo.UserDefined[k] = v
 	}
 
+	// Set content-type to default value if it is not set.
+	if _, ok := srcInfo.UserDefined["Content-Type"]; !ok {
```
Comment (harshavardhana) [inline cmd/object-handlers.go]: This is not correct @kannappanr this has to be done differently... we have to look at the incoming headers for content-type when the "REPLACE" directive is set.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#2e1ddb5d-7bc8-41b4-8cb3-c8ce29115520) |
| mint-xl.sh | :x: [more...](#fe1f6035-d01a-42f7-a451-5da18344c9ab) |
| mint-large-bucket.sh | :x: [more...](#f28e35e6-553e-4778-b6a9-cb41549f2d75) |
<a name="fe1f6035-d01a-42f7-a451-5da18344c9ab"></a>
### 6958-869ea1a/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30589
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 441c1c7f7129:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 23 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 3 seconds
(6/13) Running mc tests ... done in 10 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 36 seconds
(8/13) Running minio-go tests ... done in 45 seconds
(9/13) Running minio-java tests ... FAILED in 1 minutes and 56 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 601,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```
<a name="f28e35e6-553e-4778-b6a9-cb41549f2d75"></a>
### 6958-869ea1a/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32268
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 25ab63d4a9fa:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 1 minutes and 53 seconds
(2/13) Running aws-sdk-go tests ... done in 4 seconds
(3/13) Running aws-sdk-java tests ... done in 7 seconds
(4/13) Running aws-sdk-php tests ... done in 1 minutes and 0 seconds
(5/13) Running aws-sdk-ruby tests ... done in 32 seconds
(6/13) Running mc tests ... done in 1 minutes and 22 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 12 seconds
(8/13) Running minio-go tests ... done in 2 minutes and 15 seconds
(9/13) Running minio-java tests ... FAILED in 3 minutes and 41 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 298,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```
<a name="2e1ddb5d-7bc8-41b4-8cb3-c8ce29115520"></a>
### 6958-869ea1a/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31703
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f593d00a5828:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 9 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 4 seconds
(6/13) Running mc tests ... done in 28 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 40 seconds
(8/13) Running minio-go tests ... done in 33 seconds
(9/13) Running minio-java tests ... FAILED in 2 minutes and 11 seconds
{
  "name": "minio-java",
  "function": "copyObject(String bucketName, String objectName, String destBucketName, CopyConditions copyConditions, Map<String, String> metadata)",
  "duration": 488,
  "status": "FAIL",
  "error": "java.lang.Exception: content type differs. expected: application/javascript, got: application/octet-stream >>> [FunctionalTest.copyObject_test8(FunctionalTest.java:2314), FunctionalTest.runTests(FunctionalTest.java:2924), FunctionalTest.main(FunctionalTest.java:3027)]"
}

Executed 8 out of 13 tests successfully.
```

**Review by harshavardhana — APPROVED** [review]
LGTM we need to fix tests here @kannappanr

**Review by poornas — DISMISSED** [review]
LGTM

File: cmd/object-handlers.go [hunk 2]
```diff
@@ -643,7 +643,15 @@ func getCpObjMetadataFromHeader(ctx context.Context, r *http.Request, userMeta m
 	// if x-amz-metadata-directive says REPLACE then
 	// we extract metadata from the input headers.
 	if isMetadataReplace(r.Header) {
-		return extractMetadata(ctx, r)
+		meta, err := extractMetadata(ctx, r)
+		if err != nil {
+			return meta, err
+		}
+		// Set content-type to default value if it is not set.
+		if _, ok := meta["Content-Type"]; !ok {
```
Comment (harshavardhana) [inline cmd/object-handlers.go]: move this check inside `extractMetadata` and `Content-Type` is actually lower case here @kannappanr that's why its not working.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/object-handlers.go (hunk 2 above)
Comment (kannappanr) [reply cmd/object-handlers.go]: done

**Review by harshavardhana — APPROVED** [review]

**Review by poornas — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Pull request merged**